### PR TITLE
Report on demand

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
We've discussed this one, it's an attempt to let users say "print out a report now to this location." Still needs some work (a C API, some convenience functions), but it's worth having here so that we have a real discussion on it. I'll definitely be backing out (for example) the test changes, but that exists so the code can be played with.

@ajkunen: right now we support accepting an ostream or a FILE* as places to write a report. Is there anything else we should support? Essentially the interface is

```
cali::reporting::Reporter my_reporter = create_reporter(some_ostream_or_file_pointer);
my_reporter.report();
```

### CODE REVIEW

File: include/caliper/ReportInfrastructure.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        void flush(Caliper* c, const SnapshotRecord* flush_info) {
+            m_table_writer.flush(*c, *m_output_stream);
+        }
+        public:
+        void report(){
+          m_table_writer.flush(m_cali, *m_output_stream);
+        }
+        Reporter(output_stream_type out, const std::string& attributes, const std::string& sort, const std::string& filter, Caliper c = Caliper::instance()) :
+              m_cali(c),
```
Comment (daboehme) [inline include/caliper/ReportInfrastructure.hpp]: Don't copy the Caliper instance object please! Just get a new one when needed. (In fact, I should make the copy constructor private.)

**Comment by daboehme** [comment]
Let's wait until the flush updates are merged in. BTW, I don't see the reporter class actually triggering a Caliper flush itself - is that the intention?

**Comment by DavidPoliakoff** [comment]
I was kinda going either way on that behavior, if you think it should, let me know. My thought was that I would just write out whatever Caliper had given us, but if a flush makes the report more complete, there should be a flush.

**Comment by daboehme** [comment]
Unless you explicitly call `Caliper::flush()` it only flushes at program end. That's after you call `my_reporter.report()`. So yeah, it would make the output more complete.

**Pull request merged**